plugins/Unity/Launcher/xdgwatcher.cpp

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (C) 2019 UBports project.
+ * Author(s): Marius Gripsgard <marius@ubports.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "xdgwatcher.h"
+#include <QStandardPaths>
+#include <QDebug>
+#include <QDir>
+#include <QFileInfo>
+#include <QFile>
+#include <QTextStream>
+
+XdgWatcher::XdgWatcher(QObject* parent)
+    : QObject(parent),
+      m_watcher(new QFileSystemWatcher(this))
+{
+    connect(m_watcher, &QFileSystemWatcher::directoryChanged, this, &XdgWatcher::onDirectoryChanged);
+    connect(m_watcher, &QFileSystemWatcher::fileChanged, this, &XdgWatcher::onFileChanged);
+
+    auto paths = QStandardPaths::standardLocations(QStandardPaths::ApplicationsLocation);
+    for (auto &path: paths) {
+        auto qdir = QDir(path);
+        if (!qdir.exists()) {
+            continue;
+        }
+
+        // Add the path itself to watch for newly added apps
+        m_watcher->addPath(path);
+
+        // Add watcher for eatch app to watch for changes
+        for (auto &file: qdir.entryInfoList(QDir::Files)) {
```

Comment:
move the call to `entryInfoList()` outside of the `for` loop, and save the result into a const variable.

Reason: "When used with a (non-const) object that has copy-on-write semantics, the range-based for loop may trigger a deep copy by (implicitly) calling the non-const begin() member function. If that is undesirable (for instance because the loop is not actually modifying the object), std::as_const can be used"